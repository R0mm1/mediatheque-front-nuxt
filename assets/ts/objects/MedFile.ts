export default class MedFile {
  file: false | File | string = false
  name: string = ''
  id: string|null = null
  isNew: boolean = false

  setFile (file: File|string) {
    this.file = file
    return this
  }

  setName (name: string) {
    this.name = name
    return this
  }

  setId (id: string) {
    this.id = id
    return this
  }

  setIsNew (isNew: boolean) {
    this.isNew = isNew
    return this
  }
}
