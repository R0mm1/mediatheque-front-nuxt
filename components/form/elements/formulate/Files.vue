<template>
  <div
    :class="`formulate-input-element formulate-input-element--${context.type}`"
    :data-type="context.type"
  >
    <SimpleList :elements="listElements" :actions="listRowActions" />

    <div class="files_buttons">
      <MedInputButton
        v-if="!isFileLoading && !maxFilesReached"
        :button-descriptor="buttonAddDescriptor"
        @click.native="displayFileChooser"
      />
    </div>

    <Loader v-if="isFileLoading" class="loader" type="s" />

    <input ref="inputFile" type="file" :name="name" @change="addFile">
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator'
import SimpleList, { Element, Action } from '@/components/widgets/SimpleList.vue'
import MedInputButton from '@/components/form/elements/MedInputButton.vue'
import ButtonDescriptor from '@/assets/ts/form/ButtonDescriptor'
import Loader from '@/components/widgets/Loader.vue'
import MedFile from '~/assets/ts/objects/MedFile'

@Component({
  components: { Loader, MedInputButton, SimpleList }
})
export default class Files extends Vue {
  @Prop({ type: Object, required: true }) context!: any;

  @Prop({ default: '', type: String })name!:string;
  @Prop({ default: null, type: Number })maxFiles!:number|null;
  @Prop({ default: null, type: Function })downloadAction!:null | ((bookId: string)=>any)
  @Prop({ type: Array, required: true })files!:any[];
  @Prop({ type: Function, required: true }) onFileAdded!:((file: MedFile)=>any)
  @Prop({ type: Function })onFileRemoved?:(()=>any)

  isFileLoading: boolean = false

  displayFileChooser () {
    if (!this.maxFilesReached) {
      (this.$refs.inputFile as HTMLInputElement).click()
    }
  }

  addFile (e: Event) {
    const input = e.target as HTMLInputElement
    if (input.files === null) {
      return
    }

    this.isFileLoading = true
    const file = new MedFile()
      .setFile(input.files[0])
      .setName(input.files[0].name)
      .setIsNew(true)
    Promise.resolve(this.onFileAdded(file))
      .then(() => {
        this.isFileLoading = false
      })
  }

  downloadFile (fileId: string) {
    if (typeof this.downloadAction === 'function') {
      this.downloadAction(fileId)
    }
  }

  get maxFilesReached () {
    if (this.maxFiles === null) {
      return false
    }
    return Object.keys(this.files).length >= this.maxFiles
  }

  get listElements () {
    return this.files.map((medFile) => {
      return new Element(medFile.id, medFile.name, medFile)
    })
  }

  get listRowActions () {
    const actions = []

    const getButtonDescriptor = (name: string, faIcon: string) => {
      return (new ButtonDescriptor(name)).setFaIcon(faIcon)
    }

    if (typeof this.onFileRemoved === 'function') {
      actions.push(new Action(getButtonDescriptor('fileRemove', 'fas fa-trash-alt'), this.onFileRemoved))
    }
    if (typeof this.downloadAction === 'function') {
      actions.push(new Action(getButtonDescriptor('fileDownload', 'fas fa-file-download'), this.downloadFile))
    }
    return actions
  }

  get buttonAddDescriptor () {
    return new ButtonDescriptor('add', 'Ajouter')
  }
}
</script>

<style scoped>
input[type='file'] {
  display: none;
}

.files_buttons .form_element_button {
  float: right;
}
</style>

<style lang="scss">
.form_element_files {
  > .element_label {
    line-height: 2rem;
  }

  > .element_content {
    margin-top: 5px;
  }

  .form_element_button2 .element_content {
    display: flex;
    justify-content: flex-end;
  }
}
</style>
