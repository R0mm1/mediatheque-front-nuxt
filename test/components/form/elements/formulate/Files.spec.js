import 'reflect-metadata'
import { shallowMount } from '@vue/test-utils'
import Files from '@/components/form/elements/formulate/Files'
// import sinon from 'sinon'
import MedFile from '~/assets/ts/objects/MedFile'

describe('Files', () => {
  const wrapperFullProps = shallowMount(Files, {
    propsData: {
      context: {
        type: 'files'
      },
      name: 'electronicBook',
      maxFiles: 1,
      downloadAction: () => null,
      files: [
        new MedFile().setFile('myAwesomeBook.epub').setName('My Awesome Book').setId('123')
      ],
      onFileAdded: () => null,
      onFileRemoved: () => null
    }
  })
  const wrapperRequiredProps = shallowMount(Files, {
    propsData: {
      context: {
        type: 'files'
      },
      files: [
        new MedFile().setFile('myAwesomeBook.epub').setName('My Awesome Book').setId('123')
      ],
      onFileAdded: () => null
    }
  })

  test('is a Vue instance when specifying all props', () => {
    expect(wrapperFullProps.vm).toBeTruthy()
  })

  test('is a Vue instance when specifying only required props', () => {
    expect(wrapperRequiredProps.vm).toBeTruthy()
  })

  test('is rendering correctly when specifying all props', () => {
    expect(wrapperFullProps.element).toMatchSnapshot()
  })

  test('is rendering correctly when specifying only required props', () => {
    expect(wrapperRequiredProps.element).toMatchSnapshot()
  })

  test('is triggering click on input file when clicking add button', () => {
    const wrapper = shallowMount(Files, {
      propsData: {
        context: {
          type: 'files'
        },
        files: [],
        onFileAdded: () => null
      }
    })

    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.element).toMatchSnapshot()

    const buttonAddWrapper = wrapper.find('.files_buttons medinputbutton-stub')
    expect(buttonAddWrapper.exists()).toBeTruthy()

    const fileInputWrapper = wrapper.find('input[type="file"]')
    expect(fileInputWrapper.exists()).toBeTruthy()
    const clickInputSpy = jest.spyOn(fileInputWrapper.element, 'click')

    buttonAddWrapper.trigger('click')
    expect(clickInputSpy).toHaveBeenCalledTimes(1)
  })

  test('is calling the given onFileAdded callback when a file is added', () => {
    // const onFileAddedSpy = sinon.spy()
    // const wrapper = shallowMount(Files, {
    //   propsData: {
    //     context: {
    //       type: 'files'
    //     },
    //     files: [],
    //     onFileAdded: onFileAddedSpy
    //   }
    // })
    //
    // const fileInputWrapper = wrapper.find('input[type="file"]')
    // expect(fileInputWrapper.exists()).toBeTruthy()
    // console.log(fileInputWrapper.element.files)
    //
    // const file = new File([], 'myFile.epub')
    // const dataTransfer = new ClipboardEvent('').clipboardData || new DataTransfer()
    // dataTransfer.items.add(file)
    //
    // fileInputWrapper.element.files = dataTransfer.files
    // fileInputWrapper.trigger('change')
    // expect(onFileAddedSpy.calledOnce)
  })
})
