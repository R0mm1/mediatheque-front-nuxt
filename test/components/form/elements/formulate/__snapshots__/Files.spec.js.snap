// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Files is rendering correctly when specifying all props 1`] = `
<div
  class="formulate-input-element formulate-input-element--files"
  data-type="files"
>
  <simplelist-stub
    actions="[object Object],[object Object]"
    elements="[object Object]"
  />
   
  <div
    class="files_buttons"
  >
    <!---->
  </div>
   
  <!---->
   
  <input
    name="electronicBook"
    type="file"
  />
</div>
`;

exports[`Files is rendering correctly when specifying only required props 1`] = `
<div
  class="formulate-input-element formulate-input-element--files"
  data-type="files"
>
  <simplelist-stub
    actions=""
    elements="[object Object]"
  />
   
  <div
    class="files_buttons"
  >
    <medinputbutton-stub
      buttondescriptor="[object Object]"
    />
  </div>
   
  <!---->
   
  <input
    name=""
    type="file"
  />
</div>
`;

exports[`Files is triggering click on input file when clicking add button 1`] = `
<div
  class="formulate-input-element formulate-input-element--files"
  data-type="files"
>
  <simplelist-stub
    actions=""
    elements=""
  />
   
  <div
    class="files_buttons"
  >
    <medinputbutton-stub
      buttondescriptor="[object Object]"
    />
  </div>
   
  <!---->
   
  <input
    name=""
    type="file"
  />
</div>
`;
